### PR TITLE
Fix User Timing oddities with Suspense, pure, and lazy

### PR DESCRIPTION
**Pull request by gaearon**
This is for User Timings perf interation. We've been neglecting testing it when adding new features — let's try to keep it working in the future.

* `pure` components now show their name instead of `Unknown`
* There was a naming mismatch because `lazy` components began work as `Unknown` but completed with real name. I fix this by canceling immediately and restarting right after they get the right tag. Seems a bit odd but probably okay for this corner case?
* Suspense printed error boundary message which seemed misleading.

### CODE REVIEW

File: packages/shared/getComponentName.js [hunk 1]
```diff
@@ -63,13 +64,23 @@ function getComponentName(type: mixed): string | null {
         return 'Context.Consumer';
       case REACT_PROVIDER_TYPE:
         return 'Context.Provider';
-      case REACT_FORWARD_REF_TYPE:
+      case REACT_FORWARD_REF_TYPE: {
         const renderFn = (type.render: any);
         const functionName = renderFn.displayName || renderFn.name || '';
         return (
           (type: any).displayName ||
           (functionName !== '' ? `ForwardRef(${functionName})` : 'ForwardRef')
```
Comment (sophiebits) [inline packages/shared/getComponentName.js]: should we be consistent?

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 2]
```diff
@@ -719,11 +719,15 @@ function mountIndeterminateComponent(
     Component !== null &&
     typeof Component.then === 'function'
   ) {
+    // We can't start a User Timing measurement with correct label yet.
+    // Cancel and resume right after we know the tag.
+    cancelWorkTimer(workInProgress);
```
Comment (sophiebits) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: does this make two segments in the flame chart?

File: packages/react-reconciler/src/ReactDebugFiberPerf.js [hunk 3]
```diff
@@ -315,7 +316,10 @@ export function stopFailedWorkTimer(fiber: Fiber): void {
       return;
     }
     fiber._debugIsCurrentlyTiming = false;
-    const warning = 'An error was thrown inside this error boundary';
+    const warning =
+      fiber.tag === SuspenseComponent
+        ? 'Rendering has been suspended'
```
Comment (sophiebits) [inline packages/react-reconciler/src/ReactDebugFiberPerf.js]: Rendering was suspended?

File: packages/shared/getComponentName.js (hunk 1 above)
Comment (gaearon) [reply packages/shared/getComponentName.js]: Maybe.. I figured this one would be common and might be annoying to see it everywhere. Or maybe not?

File: packages/react-reconciler/src/ReactFiberBeginWork.js (hunk 2 above)
Comment (gaearon) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: No, it just discards the in-progress segment.

File: packages/react-reconciler/src/ReactDebugFiberPerf.js (hunk 3 above)
Comment (gaearon) [reply packages/react-reconciler/src/ReactDebugFiberPerf.js]: sure

File: packages/shared/getComponentName.js (hunk 1 above)
Comment (gaearon) [reply packages/shared/getComponentName.js]: I guess it could be helpful to see that a pure component re-rendered

**Comment by gaearon** [comment]
Addressed review

**Pull request merged**